cobc/codegen.c

```diff
@@ -13679,6 +13930,45 @@ output_header (const char *locbuff, const struct cb_program *cp)
 	}
 }
 
+static void
+output_cob_prof_data ( struct cb_program * program )
+{
+	if (cb_flag_prof) {
+		struct cb_procedure_list *l;
+		char sep = ' ';
+
+		output_local ("/* cob_prof data */\n\n");
+
+		output_local ("static const int nprocedures = %d;\n",
+			      program->procedure_list_len);
+		output_local ("static struct cob_prof_procedure prof_procedures[%d] = {\n",
+			      program->procedure_list_len);
+		sep = ' ';
+		for (l = program->procedure_list; l; l=l->next) {
+			output_local ("  %c { \"%s\", \"%s\", %d,  %d, %d }\n",
+				      sep,
+				      l->proc.text,
+				      l->proc.file,
+				      l->proc.line,
+				      l->proc.section,
+				      l->proc.kind
+				);
+			sep = ',';
+		}
+		output_local ("};\n");
+
+		output_local ("static int cob_prof_fallthrough_entry = 0;\n");
```

Comment:
This one, which may be quite important (and the missing test coverage of the new code) seems to be the only "open" issue(s).

Reply:
I *think* there is no issue here. The flag `cob_prof_fallthrough_entry` is just used very locally to skip a profiling call when we fall through an ENTRY instead of calling it.

Here is the generated code for an ENTRY:
```
  cob_prof_fallthrough_entry = 1;
l_5:;
  if (!cob_prof_fallthrough_entry)  {
    cob_prof_use_paragraph_entry (prof_info, 2, 4);
  }
  cob_prof_fallthrough_entry = 0;
```

In case of fall-through, we have the following (uninterrupted) sequence :
- set `cob_prof_fallthrough_entry` to 1
- test `cob_prof_fallthrough_entry` => it is != 0, so do nothing
- set `cob_prof_fallthrough_entry` back to 0

If we instead CALL an ENTRY, we have :
- dispatch / goto
- test `cob_prof_fallthrough_entry` => it **is** = 0, so call `cob_prof_use_paragraph_entry`
- set `cob_prof_fallthrough_entry` back to 0 (possibly redundant)

The important part is that we set `cob_prof_fallthrough_entry` to 1 **only** when falling-through, and no CALL is possible between this and setting back `cob_prof_fallthrough_entry` to 0.

So I believe we don't need to do anything - except maybe commenting about that. Or, if you feel uncomfortable with this, I can just make the variable local...

Reply:
Thank you for that, it is now clear and I agree, that a static variable is fine here.

I _think_ this would be useful for other cases as well, so want to suggest to:
* rename the variable to `fallthrough_label`
* set it only to zero in the ´else` case

Question: Do you want to have a different profiling if a section/paragraph is `PERFORM`ed / target of `GOTO` than to "fall-trough"? If yes this may also be done in a follow-up PR.

Reply:
> Question: Do you want to have a different profiling if a section/paragraph is `PERFORM`ed / target of `GOTO` than to "fall-trough"? If yes this may also be done in a follow-up PR.

Might be a good idea indeed.

Reply:
`cob_prof_fallthrough_label` is already better, but as this is a local field we don't need the `cob_` prefix, and as we (likely) use that for other things than profiling in the future, just going with ´fallthrough_label` seems useful.